### ISSUE TITLE
Should Chapel enums support non-param values?

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#10115]
In the old world, Chapel enums were essentially treated as a collection of associated param ints given the implicit conversions they supported and the fact that every enum symbol had an associated integer value.  In the current world, Chapel enums no longer necessarily have integer values and no longer support implicit conversions.  This makes me wonder whether we could / should relax the requirement that the initializing expressions be param integer values.  I'm not sure what it's buying us anymore.

**Comment by gbtitus** [comment]
Could an enum type define a mix of symbols whose associated values were a mixture of params and non-params?  Or would the presence of at least one non-param associated value essentially make the whole type a non-param one?  I'm trying to think about the implications for users declaring param or non-param variables of such enum types.  Obviously we'd want the compiler to bark if it saw code declaring a param enum var but initializing it with an enum value of that type, whose associated value wasn't param.  I just don't know how hard we want the compiler to have to work to do that.  I guess there's also the question of how hard we want users to have to work to reason about such enum values.

**Comment by gbtitus** [comment]
There's an interaction between this issue and #10139, I think.  Suppose we allow enum associated values to be execution-time values.  Then the following would be legal, presumably:

```chapel
enum epoch { past = 0, present = getCurrentTime():int, future = 999999999999 };
```

`epoch.present` probably won't be the same from run to run.  So if you wanted to be able to write a variable of that enum type in one run of a program and read it back in another, in order for it to make sense when read back the value stored in the file would have to be something that didn't depend on the associated value.

**Comment by bradcray** [comment]
> So if you wanted to be able to write a variable of that enum type in one run of a program and read it back in another, in order for it to make sense when read back the value stored in the file would have to be something that didn't depend on the associated value.

Good point.  Note that other changes to an enum (such as reordering its symbols or changing the param values of some of them, e.g., via a config) could also cause problems in any binary I/O scheme that we're discussing here, but your example has the benefit of not requiring source changes or changes to compilation options to make the point.

**Comment by gbtitus** [comment]
> Note that other changes to an enum (such as reordering its symbols or changing the param values of some of them, e.g., via a config) could also cause problems in any binary I/O scheme that we're discussing here ...

Right, although I put the comment here, my chain of thought on this actually started with #10139 and considering the requirements for run-to-run syntactic and semantic stability each alternative would impose.  In order to allow for reordering of the enum symbols within a type the binary file storage form would need to be the actual symbol name or something that could be uniquely mapped to that, and I'm sure we don't want to go there.  :-)

**Comment by bradcray** [comment]
Reviewing this issue as part of #9221, I'll say that it seems conceivable to me that we could support it, but that it's also easy work to put off until someone wants/needs it.
